### ISSUE TITLE
webhook: on_message_publish can not work well

### ISSUE THREAD

**Issue by RichardChengshaojin** [emqx/emqx#3583]
**Please describe your problem in detail, if necessary, you can upload the log file through the attachment**:
- OS: Linux version 5.6.2-1.el7.elrepo.x86_64 (mockbuild@Build64R7) (gcc version 4.8.5 20150623 (Red Hat 4.8.5-39) (GCC)) #1 SMP Thu Apr 2 10:55:54 EDT 2020
- EMQX version: 3.0.1 package url: https://www.emqx.io/downloads/broker/v3.0.1/emqx-centos7-v3.0.1.x86_64.rpm

- Question:
Event, on_client_connected on_client_disconnected on_client_subscribe on_client_unsubscribe
on_session_created  on_session_subscribed on_session_unsubscribed  on_session_terminated, can work well and my web server can receive hook message successfully.

but when I enable 

`web.hook.rule.message.publish.1      = {"action": "on_message_publish"}`

Something goes wrong during message publishing, it seems like I can not publish message to the broker. The broker log shows below:

```log
2020-07-11 17:42:39.629 [debug] emqx_test WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"session_created","client_id":"emqx_test","username":"radioServer"}
2020-07-11 17:42:39.635 [debug] emqx_test@10.150.123.214:1773 WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"client_connected","client_id":"emqx_test","username":"radioServer","conn_ack":0}
2020-07-11 17:42:39.640 [debug] emqx_test@10.150.123.214:1773 SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)
2020-07-11 17:42:39.640 [debug] emqx_test@10.150.123.214:1773 Keepalive at the interval of 4
2020-07-11 17:42:39.643 [debug] emqx_test@10.150.123.214:1773 RECV <<130,16,0,1,0,11,116,101,115,116,116,111,112,105,99,47,50,1>>
2020-07-11 17:42:39.643 [debug] emqx_test@10.150.123.214:1773 RECV SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicFilters=[{<<"testtopic/2">>,
                           #{nl => 0,qos => 1,rap => 0,rc => 0,rh => 0}}])
2020-07-11 17:42:39.644 [debug] emqx_test@10.150.123.214:1773 WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"client_subscribe","client_id":"emqx_test","username":"radioServer","topic":"testtopic/2","opts":{"nl":0,"qos":1,"rap":0,"rc":0,"rh":0}}
2020-07-11 17:42:39.648 [debug] emqx_test WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"session_subscribed","client_id":"emqx_test","topic":"testtopic/2","opts":{"first":true,"nl":0,"qos":1,"rap":0,"rc":0,"rh":0}}
2020-07-11 17:42:39.650 [debug] emqx_test@10.150.123.214:1773 SEND SUBACK(Q0, R0, D0, PacketId=1, ReasonCodes=[1])
2020-07-11 17:42:39.653 [debug] emqx_test@10.150.123.214:1773 RECV <<52,26,0,11,116,101,115,116,116,111,112,105,99,47,50,0,2,72,101,108,108,
       111,32,87,111,114,108,100>>
2020-07-11 17:42:39.653 [debug] emqx_test@10.150.123.214:1773 RECV PUBLISH(Q2, R0, D0, Topic=testtopic/2, PacketId=2, Payload=<<"Hello World">>)
2020-07-11 17:42:39.654 [info] emqx_test@10.150.123.214:1773 PUBLISH to testtopic/2: <<"Hello World">>
```

And client code `client.publish(pubTopic, message);` throws Exception.

What I want:

- Where Can I find the plugin(emqx_web_hook)'s log ?
- And Could you give some hints about this problem ?

Many thanks.

**Comment by HJianBo** [comment]
That's is all logs for emqx and webhook plugin.

First of all, after updated `emqx_web_hook.conf`, please reload emqx_web_hook plugin first.

The `[debug] emqx_test WebHook: Url:"http://10.150.123.214:8080/emqx/hook ..` log must be printed, if the WebHook wants to send an HTTP Request to your server.

So, do your broker blocked somewhere? What's your plugins list started?

**Comment by RichardChengshaojin** [comment]
@HJianBo 

I restart the emqx broker and reload plugin through dashboard.

Here's my server code, do nothing bug log message:

```java
@Slf4j
@RestController
public class EmqHookController {

    @ResponseStatus(HttpStatus.OK)
    @RequestMapping(value = "emqx/hook")
    public void hook(@RequestBody String body) {
        log.info("received => {}", body);
    }
}
```

And Here are my plugin list:

```log
[root@devops1722086 bin]# ./emqx_ctl plugins list
Plugin(emqx_auth_clientid, version=3.0, description=EMQ X Authentication with ClientId/Password, active=false)
Plugin(emqx_auth_http, version=v3.0.1, description=EMQ X Authentication/ACL with HTTP API, active=false)
Plugin(emqx_auth_jwt, version=3.0, description=EMQ X Authentication with JWT, active=false)
Plugin(emqx_auth_ldap, version=3.0, description=EMQ X Authentication/ACL with LDAP, active=false)
Plugin(emqx_auth_mongo, version=3.0, description=EMQ X Authentication/ACL with MongoDB, active=false)
Plugin(emqx_auth_mysql, version=3.0, description=EMQ X Authentication/ACL with MySQL, active=false)
Plugin(emqx_auth_pgsql, version=3.0, description=EMQ X Authentication/ACL with PostgreSQL, active=false)
Plugin(emqx_auth_redis, version=3.0, description=EMQ X Authentication/ACL with Redis, active=false)
Plugin(emqx_auth_username, version=3.0, description=EMQ X Authentication with Username and Password, active=false)
Plugin(emqx_coap, version=3.0, description=EMQ X CoAP Gateway, active=false)
Plugin(emqx_dashboard, version=3.0, description=EMQ X Web Dashboard, active=true)
Plugin(emqx_delayed_publish, version=3.0, description=EMQ X Delayed Publish, active=false)
Plugin(emqx_lua_hook, version=3.0, description=EMQ X Lua Hooks, active=false)
Plugin(emqx_lwm2m, version=3.0, description=EMQ X LwM2M Gateway, active=false)
Plugin(emqx_management, version=3.0, description=EMQ X Management API and CLI, active=true)
Plugin(emqx_plugin_template, version=3.0, description=EMQ X Plugin Template, active=false)
Plugin(emqx_recon, version=3.0, description=EMQ X Recon Plugin, active=true)
Plugin(emqx_reloader, version=3.0, description=EMQ X Reloader Plugin, active=false)
Plugin(emqx_retainer, version=3.0, description=EMQ X Retainer, active=true)
Plugin(emqx_sn, version=3.0, description=EMQ X MQTT SN Plugin, active=false)
Plugin(emqx_statsd, version=3.0, description=Statsd for EMQ X, active=false)
Plugin(emqx_stomp, version=3.0, description=EMQ X Stomp Protocol Plugin, active=false)
Plugin(emqx_web_hook, version=3.0, description=EMQ X Webhook Plugin, active=true)
```

**Comment by HJianBo** [comment]
It seems correct.

Maybe you can try to debug it:
1.  Use `bin/emqx console` to start broker
2. Typing `dbg:tracer(),dbg:p(all,call), dbg:tp(emqx_web_hook, on_message_publish,x).` and `Enter` to debug emqx_web_hook's function 
3. Try it again and paste the output to there

**Comment by RichardChengshaojin** [comment]
> It seems correct.
> 
> Maybe you can try to debug it:
> 
> 1. Use `bin/emqx console` to start broker
> 2. Typing `dbg:tracer(),dbg:p(all,call), dbg:tp(emqx_web_hook, on_message_publish,x).` and `Enter` to debug emqx_web_hook's function
> 3. Try it again and paste the output to there

@HJianBo 

I used console to start EMQX broker, and typed:

`dbg:tracer(),dbg:p(all,call), dbg:tp(emqx_web_hook, on_message_publish,x)`

but got nothing:

```java
2020-07-13 09:59:00.602 [info] Load emqx_mod_presence module successfully.
Start http:management listener on 8080 successfully.
2020-07-13 09:59:01.066 [info] Started Apps: [minirest,emqx_management]
2020-07-13 09:59:01.067 [info] Load plugin emqx_management successfully
2020-07-13 09:59:01.069 [info] Started Apps: [recon,emqx_recon]
2020-07-13 09:59:01.069 [info] Load plugin emqx_recon successfully
2020-07-13 09:59:01.073 [info] Started Apps: [emqx_retainer]
2020-07-13 09:59:01.073 [info] Load plugin emqx_retainer successfully
Start http:dashboard listener on 18083 successfully.
2020-07-13 09:59:01.323 [info] Started Apps: [emqx_dashboard]
2020-07-13 09:59:01.323 [info] Load plugin emqx_dashboard successfully
2020-07-13 09:59:01.327 [info] Started Apps: [emqx_web_hook]
2020-07-13 09:59:01.328 [info] Load plugin emqx_web_hook successfully
Start mqtt:tcp listener on 127.0.0.1:11883 successfully.
Start mqtt:tcp listener on 0.0.0.0:1883 successfully.
Start mqtt:ws listener on 0.0.0.0:8083 successfully.
Start mqtt:ssl listener on 0.0.0.0:8883 successfully.
Start mqtt:wss listener on 0.0.0.0:8084 successfully.
EMQ X Broker v3.0.1 is running now!
Eshell V10.2.1  (abort with ^G)
(emqx@127.0.0.1)1> 
(emqx@127.0.0.1)1> dbg:tracer(),dbg:p(all,call), dbg:tp(emqx_web_hook, on_message_publish,x) 
(emqx@127.0.0.1)1> 
```

**Comment by HJianBo** [comment]
emmm... you missed the `.`  character in the end

**Comment by RichardChengshaojin** [comment]
Aha I know nothing about Erlang~ here is the output: 

```java
(emqx@127.0.0.1)1> dbg:tracer(),dbg:p(all,call),dbg:tp(emqx_web_hook,on_message_publish,x).
{ok,[{matched,'emqx@127.0.0.1',1},{saved,x}]}
(emqx@127.0.0.1)2> (<0.1476.0>) call emqx_web_hook:on_message_publish({message,<<0,5,170,73,38,198,142,0,244,64,0,0,5,196,0,42>>,
         0,emqx_sys,
         #{dup => false,sys => true},
         #{},<<"$SYS/brokers/emqx@127.0.0.1/uptime">>,<<"22 seconds">>,
         {1594,606043,434500}},{undefined}) 
(<0.1476.0>) returned from emqx_web_hook:on_message_publish/2 -> {ok,
                                                                  {message,
                                                                   <<0,5,170,
                                                                     73,38,198,
                                                                     142,0,244,
                                                                     64,0,0,5,
                                                                     196,0,42>>,
                                                                   0,emqx_sys,
                                                                   #{dup =>
                                                                      false,
                                                                     sys =>
                                                                      true},
                                                                   #{},
                                                                   <<"$SYS/brokers/emqx@127.0.0.1/uptime">>,
                                                                   <<"22 seconds">>,
                                                                   {1594,
                                                                    606043,
                                                                    434500}}} 
(<0.1476.0>) call emqx_web_hook:on_message_publish({message,<<0,5,170,73,38,198,142,202,244,64,0,0,5,196,0,43>>,
         0,emqx_sys,
         #{dup => false,sys => true},
         #{},<<"$SYS/brokers/emqx@127.0.0.1/datatype">>,
         <<"2020-07-13 10:07:23">>,
         {1594,606043,434699}},{undefined})
(<0.1476.0>) returned from emqx_web_hook:on_message_publish/2 -> {ok,
                                                                  {message,
                                                                   <<0,5,170,
                                                                     73,38,198,
                                                                     142,202,
                                                                     244,64,0,
                                                                     0,5,196,0,
                                                                     43>>,
                                                                   0,emqx_sys,
                                                                   #{dup =>
                                                                      false,
                                                                     sys =>
                                                                      true},
                                                                   #{},
                                                                   <<"$SYS/brokers/emqx@127.0.0.1/datatype">>,
                                                                   <<"2020-07-13 10:07:23">>,
                                                                   {1594,
                                                                    606043,
                                                                    434699}}}

```

**Comment by HJianBo** [comment]
Sorry, This command has more output here. Plz use `dbg:tracer(),dbg:p(all,call), dbg:tpl(emqx_web_hook, send_http_request,x).` instead.  Then Try to send some messages from your MQTT Client and paste the output to here

**Comment by RichardChengshaojin** [comment]
I typed  `dbg:tracer(),dbg:p(all,call),dbg:tpl(emqx_web_hook,send_http_request,x).` and then sent a message to the broker.
All output is shown as below:

```java
(emqx@127.0.0.1)1> dbg:tracer(),dbg:p(all,call),dbg:tpl(emqx_web_hook,send_http_request,x).
{ok,[{matched,'emqx@127.0.0.1',1},{saved,x}]}
(emqx@127.0.0.1)2> 2020-07-13 10:17:06.073 [debug] 10.150.123.214:3795 RECV <<16,100,0,4,77,81,84,84,4,194,0,5,0,9,101,109,113,120,95,116,101,115,116,
       0,11,114,97,100,105,111,83,101,114,118,101,114,0,64,102,99,51,50,102,56,
       53,49,53,48,101,51,53,53,57,54,100,50,49,50,53,49,48,54,51,51,52,49,101,
       56,48,54,97,52,101,53,49,55,56,98,99,51,49,102,52,53,57,57,51,99,101,99,
       56,55,57,57,54,57,50,102,56,101,99,102>>
2020-07-13 10:17:06.075 [debug] 10.150.123.214:3795 RECV CONNECT(Q0, R0, D0, ClientId=emqx_test, ProtoName=MQTT, ProtoVsn=4, CleanStart=true, KeepAlive=5, Username=radioServer, Password=******)
(<0.1638.0>) call emqx_web_hook:send_http_request([{action,session_created},
 {client_id,<<"emqx_test">>},
 {username,<<"radioServer">>}])
2020-07-13 10:17:06.076 [debug] emqx_test WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"session_created","client_id":"emqx_test","username":"radioServer"}
(<0.1638.0>) returned from emqx_web_hook:send_http_request/1 -> ok
(<0.1636.0>) call emqx_web_hook:send_http_request([{action,client_connected},
 {client_id,<<"emqx_test">>},
 {username,<<"radioServer">>},
 {conn_ack,0}])
2020-07-13 10:17:06.104 [debug] emqx_test@10.150.123.214:3795 WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"client_connected","client_id":"emqx_test","username":"radioServer","conn_ack":0}
(<0.1636.0>) returned from emqx_web_hook:send_http_request/1 -> ok
2020-07-13 10:17:06.112 [debug] emqx_test@10.150.123.214:3795 SEND CONNACK(Q0, R0, D0, AckFlags=0, ReasonCode=0)
2020-07-13 10:17:06.113 [debug] emqx_test@10.150.123.214:3795 Keepalive at the interval of 4
2020-07-13 10:17:06.117 [debug] emqx_test@10.150.123.214:3795 RECV <<130,16,0,1,0,11,116,101,115,116,116,111,112,105,99,47,50,1>>
2020-07-13 10:17:06.117 [debug] emqx_test@10.150.123.214:3795 RECV SUBSCRIBE(Q1, R0, D0, PacketId=1, TopicFilters=[{<<"testtopic/2">>,
                           #{nl => 0,qos => 1,rap => 0,rc => 0,rh => 0}}])
(<0.1636.0>) call emqx_web_hook:send_http_request([{action,client_subscribe},
 {client_id,<<"emqx_test">>},
 {username,<<"radioServer">>},
 {topic,<<"testtopic/2">>},
 {opts,#{nl => 0,qos => 1,rap => 0,rc => 0,rh => 0}}])
2020-07-13 10:17:06.120 [debug] emqx_test@10.150.123.214:3795 WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"client_subscribe","client_id":"emqx_test","username":"radioServer","topic":"testtopic/2","opts":{"nl":0,"qos":1,"rap":0,"rc":0,"rh":0}}
(<0.1636.0>) returned from emqx_web_hook:send_http_request/1 -> ok
(<0.1638.0>) call emqx_web_hook:send_http_request([{action,session_subscribed},
 {client_id,<<"emqx_test">>},
 {topic,<<"testtopic/2">>},
 {opts,#{first => true,nl => 0,qos => 1,rap => 0,rc => 0,rh => 0}}])
2020-07-13 10:17:06.126 [debug] emqx_test WebHook: Url:"http://10.150.123.214:8080/emqx/hook", params:{"action":"session_subscribed","client_id":"emqx_test","topic":"testtopic/2","opts":{"first":true,"nl":0,"qos":1,"rap":0,"rc":0,"rh":0}}
(<0.1638.0>) returned from emqx_web_hook:send_http_request/1 -> ok
2020-07-13 10:17:06.129 [debug] emqx_test@10.150.123.214:3795 SEND SUBACK(Q0, R0, D0, PacketId=1, ReasonCodes=[1])
2020-07-13 10:17:06.137 [debug] emqx_test@10.150.123.214:3795 RECV <<52,26,0,11,116,101,115,116,116,111,112,105,99,47,50,0,2,72,101,108,108,
       111,32,87,111,114,108,100>>
2020-07-13 10:17:06.138 [debug] emqx_test@10.150.123.214:3795 RECV PUBLISH(Q2, R0, D0, Topic=testtopic/2, PacketId=2, Payload=<<"Hello World">>)
2020-07-13 10:17:06.139 [info] emqx_test@10.150.123.214:3795 PUBLISH to testtopic/2: <<"Hello World">>
```

**Comment by RichardChengshaojin** [comment]
@HJianBo 

And I got some other output a few minutes later, I am not sure if it has anything to do with my operations above:

```java
2020-07-13T10:22:06.139939+08:00 warning: FORMATTER CRASH: {report,#{}}
2020-07-13 10:22:06.144 [debug] formatter_crashed: emqx_logger_formatter
    config: #{template =>
                  [time," [",level,"] ",
                   {client_id,
                       [{peername,
                            [client_id,"@",peername," "],
                            [client_id," "]}],
                       [{peername,[peername," "],[]}]},
                   msg,"\n"]}
    log_event: #{level => warning,
                 meta =>
                     #{client_id => <<"emqx_test">>,gl => <0.1355.0>,
                       pid => <0.1638.0>,
                       report_cb => #Fun<emqx_session.17.27769695>,
                       time => 1594606926139939},
                 msg => {report,#{}}}
    reason: {error,function_clause,
                [{emqx_logger_formatter,format_msg,
                     [{"FORMAT ERROR: ~0tp - ~0tp",
                       ["Dropped qos2 packet ~s for await_rel_timeout",[2]]},
                      unlimited,[]],
                     [{file,"src/emqx_logger_formatter.erl"},{line,143}]},
                 {emqx_logger_formatter,format,2,
                     [{file,"src/emqx_logger_formatter.erl"},{line,77}]}]}
```

**Comment by HJianBo** [comment]
It is a so weird appearance. Is it convenient for WeChat? Plz add me `Lo_Alone`

**Comment by RichardChengshaojin** [comment]
@HJianBo sure. I've sent a friend request.

**Comment by HJianBo** [comment]
Sorry, @RichardChengshaojin. We have reproduced it. 

It is a bug on `emqx-web-hook` plugin introduced at v3.0-rc.5 and fixed it at v3.1.0 first.

But we have no plan to release the `v3.0.2` to fix it. So plz upgrade the broker to `v3.1.0` or `v3.2.7` or the latest version `4.1.1`

Thanks for your feedback here  :-)
